### ISSUE TITLE
Emitted event's arguments do not preserve the reference to the original object

### ISSUE THREAD

**Issue by dumptyd** [vuejs/vue-test-utils-next#1044]
### Versions

**vue-test-utils** - 2.0.0-rc.16
**vue-jest** - 5.0.0-alpha.10
**jest** - 24.9.0

### Subject of the issue

I have a smoke test to ensure that when I emit a particular event, I emit it with the same object that was passed to the component. This test uses `expect(emittedObj).toBe(passedObj)` (Object.is) check to ensure the reference is preserved.

I'm migrating my library to Vue 3 and this test started breaking when using vtu v2 and vue-jest v5.
I ran the app in my browser to make sure this isn't a Vue issue and it isn't (reference is preserved in the real environment.)

### Steps to reproduce

```js
import { mount } from '@vue/test-utils';

const Comp = {
  template: `<div @click="$emit('custom-event', obj)">Hello, world!</div>`,
  props: {
    obj: Object
  }
};

it('emits event', async () => {
  const obj = { key: 'value' };
  const wrapper = mount(Comp, { props: { obj } });

  await wrapper.trigger('click');
  const emittedObj = wrapper.emitted('custom-event')[0][0];
  expect(emittedObj).toBe(obj);
});
```
<details>
  <summary><strong>Result (failing)</strong></summary>
  <img src="https://user-images.githubusercontent.com/16368429/139177788-7563aac4-52e2-4547-a32b-dd1e9819e0f1.png">
</details>

### Expected behaviour

Reference should be preserved. Here's the equivalent syntax for VTU v1 that works as expected:

```js
import { mount } from '@vue/test-utils'

const Comp = {
  template: `<div @click="$emit('custom-event', obj)">Hello, world!</div>`,
  props: {
    obj: Object
  }
};

it('emits event', async () => {
  const obj = { key: 'value' };
  const wrapper = mount(Comp, { propsData: { obj } });

  wrapper.trigger('click');
  const emittedObj = wrapper.emitted('custom-event')[0][0];
  expect(emittedObj).toBe(obj);
});
```

### Actual behaviour

Reference is not preserved.

### Possible Solution

`toStrictEqual` is a possible temporary alternative but that doesn't do a reference check.

**Comment by lmiller1990** [comment]
It looks like we make the props reactive [here](https://github.com/vuejs/vue-test-utils-next/blob/master/src/mount.ts#L319-L323) to support features like `setProps`. 

I guess this technically is a bug, since it's a different behavior from what Vue does.

I don't know if we would have any easy way to support this; I'm also not sure it's really advisable to rely on the object reference to be the same.. Is using `toEqual` an alternative?

**Comment by dumptyd** [comment]
I looked into this more when I was working on that specific test and came to the same conclusion as you that somewhere down the line, the object is becoming reactive and that's what causes this.

But, I also realized that this is the correct behavior and VTU v1 has the _incorrect_ behavior.

**TL;DR: I think the issue can be closed.**

---
#### Why it works in the browser
>I ran the app in my browser to make sure this isn't a Vue issue and it isn't (reference is preserved in the real environment.)

When I'm passing a prop to a child component, that specific object has already been made reactive by the time it's defined on the instance, so the manual check passes. I'll try to convey this with an example -
```vue
<template>
  <component-to-test :prop-a="obj" @event="handleEvent"></component-to-test>
</template>
<script>
  export default {
    data() {
      return {
        // already reactive by the time this function returns and the object reaches `component-to-test`
        obj: { key: 'value' }
      }
    },
    methods: {
      handleEvent(emittedObj) {
        console.log(this.obj === emittedObj); // true. because the check is being done after reactivity has been added
      }
    }
  };
</script>
```

---

#### Correct way to perform this check

As for the solution, I simulated the above by using a temporary component that renders component that needs to be tested and that did the trick.

```js
const Comp = {
  template: '<div @click="$emit(\'custom-event\', obj)">Hello, world!</div>',
  props: {
    obj: Object
  }
};

it('emits event', async() => {
  const RenderedComp = {
    template: '<comp :obj="obj"></comp>',
    components: { Comp },
    data() {
      return {
        obj: { key: 'value' }
      };
    }
  };
  const wrapper = mount(RenderedComp);

  const compWrapper = wrapper.findComponent(Comp);
  await compWrapper.trigger('click');
  const emittedObj = compWrapper.emitted('custom-event')[0][0];
  expect(emittedObj).toBe(RenderedComp.vm.obj);
});
```

**Comment by lmiller1990** [comment]
Interesting... it's really difficult to get VTU v1 and v2 to do exactly the same thing, considering the internals are so different (Vue and Test Utils).

Either way, it seems like you found a solution. I don't expect this "emitted object should be referentially identical" edge case to be a problem outside a small number of cases, so I think we can close this one.

I'm still not fully understanding the need to use `toBe` here; it seems you could use `toEqual` and not need the temporary component trick 🤔
